File: 682-baseball-game/baseball-game.py
class Solution:
    def calPoints(self, operations: List[str]) -> int:
        stack = [] # basically a list
        sum_of_scores = 0
        
        for i in range (len(operations)):
            if(operations[i]=="C"):
                stack.pop()
            elif(operations[i]=="D"):
                top = stack[-1]
                stack.append(top*2)
            elif(operations[i]=="+"):
                top = stack[-1]
                secondLast = stack[-2]
                stack.append(top + secondLast)
            else:
                stack.append(int(operations[i]))
        
        for s in stack:
            sum_of_scores += s
        
        return sum_of_scores

File: 682-baseball-game/baseball-game.java
class Solution {
    public int calPoints(String[] operations) {

        int sumOfScores = 0;
        Stack<Integer> stack = new Stack<Integer>();
        // we use a stack to keep track of valid scores 
        // and to access and update the last or second-last elements based on the operations
        
        for(int i=0; i<operations.length; i++){
            if(operations[i].equals("C")){
                stack.pop();
            }
            else if(operations[i].equals("D")){
                int top = stack.peek();
                stack.push(top * 2);
            }
            else if(operations[i].equals("+")){
                int top = stack.peek();
                int secondLast = stack.get(stack.size() - 2);
                stack.push(top + secondLast);
                // stack.size() gives the total number of elements
                // stack.get(index) retrieves the element at that index
                // stack.get(stack.size() - 2) gives the second-last element
            }
            else{
                stack.push(Integer.parseInt(operations[i]));
            }
        }
        for(Integer s : stack){
            sumOfScores += s;
        }
        return sumOfScores;
    }
}